### PR TITLE
(#67) add controller Users, change state and roles

### PR DESCRIPTION
**Pull request by reabiliti**
https://github.com/howtohireme/give-me-poc/issues/67
Добавил контроллер для пользователей, отображает всех пользователей со статусом active или disabled, можно менять статус.
default
Создал отдельную страницу с пользователями, где можно удалять или добавлять роль.
default
default

Как я понял, еще делается параллельно задача по отправке почты пользователю, пока не разбирался и не добавлял уведомление об изменении статуса пользователя.

### CODE REVIEW

File: spec/controllers/web/dashboard/users_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    context 'not signed in' do
+      before { get :index }
+
+      it 'redirect to root landing' do
+        expect(response).to redirect_to '/'
+      end
+    end
+
+    context 'not authorized' do
+      let(:user) { create(:user, :active) }
```
Comment (Mehonoshin) [inline spec/controllers/web/dashboard/users_controller_spec.rb]: you missed the `login_user` in this context.
It is weird that this spec passes. Please investigate why that happens.

File: spec/controllers/web/dashboard/users_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+      before { get :index }
+
+      it 'redirect to dashboard root' do
+        expect(response).to redirect_to dashboard_root_path
+      end
+    end
+
+    context 'staff' do
+      before { login_user(user) }
```
Comment (Mehonoshin) [inline spec/controllers/web/dashboard/users_controller_spec.rb]: Merge two before blocks into one

File: spec/controllers/web/dashboard/users_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+      it 'returns success status' do
+        expect(response.status).to eq 200
+      end
+    end
+  end
+
+  describe 'GET #show' do
+    context 'authorized' do
+      let!(:candidate) { create(:user, :screening_completed) }
+      before { login_user(user) }
```
Comment (Mehonoshin) [inline spec/controllers/web/dashboard/users_controller_spec.rb]: merge

File: spec/controllers/web/dashboard/users_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        expect(Ops::Developer::Activate).to receive(:call).with(user: candidate)
+        put :activate, params: { id: candidate.id }
+      end
+
+      it 'redirect to users list' do
+        put :activate, params: { id: candidate.id }
+        expect(response).to redirect_to dashboard_users_path
+      end
+
+      it 'flash' do
```
Comment (Mehonoshin) [inline spec/controllers/web/dashboard/users_controller_spec.rb]: `assigns success flash`

File: spec/controllers/web/dashboard/users_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    context 'authorized' do
+      let!(:candidate) { create(:user, :disable) }
+      before { login_user(user) }
+
+      it 'calls Activate operation' do
+        expect(Ops::Developer::Activate).to receive(:call).with(user: candidate)
+        put :activate, params: { id: candidate.id }
+      end
+
+      it 'redirect to users list' do
```
Comment (Mehonoshin) [inline spec/controllers/web/dashboard/users_controller_spec.rb]: redirects

File: spec/controllers/web/dashboard/users_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+      it 'redirect to dashboard root' do
+        put :disable, params: { id: candidate.id }
+        expect(response).to redirect_to dashboard_root_path
+      end
+    end
+  end
+
+  describe 'PUT #add_role' do
+    let(:role) { 'stuff' }
```
Comment (Mehonoshin) [inline spec/controllers/web/dashboard/users_controller_spec.rb]: add a blank line before `let` and `context`

File: app/operations/ops/developer/add_role.rb [hunk 7]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+module Ops
+  module Developer
+    # This add role to user
+    class AddRole < BaseOperation
```
Comment (Mehonoshin) [inline app/operations/ops/developer/add_role.rb]: Lets call it `AssignRole`

File: spec/factories/users.rb [hunk 8]
```diff
@@ -18,6 +18,10 @@
       state 'active'
     end
 
+    trait :disable do
```
Comment (Mehonoshin) [inline spec/factories/users.rb]: `disabled`

File: spec/factories/users.rb [hunk 9]
```diff
@@ -28,6 +32,12 @@
       end
     end
 
+    trait :developer do
```
Comment (Mehonoshin) [inline spec/factories/users.rb]: actually, developer role is assigned by default, via callback in the model.
So you can leave this trait as empty.

File: app/controllers/web/dashboard/users_controller.rb [hunk 10]
```diff
@@ -0,0 +1,52 @@
+# frozen_string_literal: true
+
+module Web
+  module Dashboard
+    # Users management logic
+    class UsersController < BaseController
+      before_action :user_find, except: :index
+      before_action :authorize_staff!
```
[the remark this answers is not in the document]
Comment (reabiliti) [reply app/controllers/web/dashboard/users_controller.rb]: fix

File: app/models/user.rb [hunk 11]
```diff
@@ -25,6 +25,8 @@ class User < ApplicationRecord
 
   accepts_nested_attributes_for :authentications
 
+  scope :active_disabled, -> { where(state: %w[active disabled]).order(email: :asc) }
```
[the remark this answers is not in the document]
Comment (reabiliti) [reply app/models/user.rb]: fix

File: app/controllers/web/dashboard/users_controller.rb [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+      def add_role
+        Ops::Developer::AddRole.call(user: @user, role: params[:role])
+        redirect_to dashboard_user_path(@user),
+                    flash: { success: "#{params[:role]} #{t('dashboard.users.notices.add_role')}" }
+      end
+
+      def remove_role
+        Ops::Developer::RemoveRole.call(user: @user, role: params[:role], size: @user.roles.size)
+        redirect_to dashboard_user_path(@user)
```
[the remark this answers is not in the document]
Comment (reabiliti) [reply app/controllers/web/dashboard/users_controller.rb]: fix

File: app/operations/ops/developer/add_role.rb [hunk 13]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+module Ops
+  module Developer
+    # This add role to user
```
[the remark this answers is not in the document]
Comment (reabiliti) [reply app/operations/ops/developer/add_role.rb]: fix

File: app/views/web/dashboard/users/index.html.haml [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+          %td
+            - user.roles.each_with_index do |role, index|
+              - if index.zero?
+                = role.name
+              - else
+                %br/
+                = role.name
+          %td
+            - if user.active?
+              = link_to user.state, disable_dashboard_user_path(user), method: :put
```
[the remark this answers is not in the document]
Comment (reabiliti) [reply app/views/web/dashboard/users/index.html.haml]: fix

File: app/views/web/dashboard/users/show.html.haml [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+          - unless @user.roles_name.include?(role)
+            %br/
+            = link_to "add role #{role}", add_role_dashboard_user_path(@user, role: role), method: :put
+    %tr
+      %td
+        %b
+          Task assignments
+      %td
+        - @user.test_task_assignments.each_with_index do |test_task_assignment, index|
+          - if index.zero?
```
[the remark this answers is not in the document]
Comment (reabiliti) [reply app/views/web/dashboard/users/show.html.haml]: fix

File: config/routes.rb [hunk 16]
```diff
@@ -30,6 +30,14 @@
           put :reject
         end
       end
+      resources :users, only: %i[index show] do
+        member do
+          put :activate
+          put :disable
```
[the remark this answers is not in the document]
Comment (reabiliti) [reply config/routes.rb]: is used:
flash: { success: "#{params[:role]} #{t('dashboard.users.notices.add_role')}" }
you get the following message:
default

File: config/locales/dashboard/en.yml [hunk 17]
```diff
@@ -7,3 +7,13 @@ en:
       reject: Reject
       activate: Activate
       confirm_rejection: Do you really want to reject this candidate?
+    users:
+      access:
+        deny: The user does not have the role of stuff
+      notices:
+        activated: The user was activated
+        disabled: The user was disabled
+        add_role: role was added
```
[the remark this answers is not in the document]
Comment (reabiliti) [reply config/locales/dashboard/en.yml]: is used:
flash: { success: "#{params[:role]} #{t('dashboard.users.notices.add_role')}" }
you get the following message:
default

[the remark this answers is not in the document]
Comment (Mehonoshin) [reply config/locales/dashboard/en.yml]: ok then, but please capitalize role name in the flash.
All sentences should start with capital letter.

**Comment by Mehonoshin** [comment]
Also please add several users through `db/seeds.rb` for development purposes.

File: spec/controllers/web/dashboard/users_controller_spec.rb (hunk 1 above)
Comment (reabiliti) [reply spec/controllers/web/dashboard/users_controller_spec.rb]: I did following the example:
describe Web::Dashboard::TestTaskAssignmentsController do
  let(:user) { create(:user, :staff) }

  describe 'GET #index' do
    context 'not signed in' do
      before { get :index }

      it 'redirects to root landing' do
        expect(response).to redirect_to '/'
      end
    end

    context 'not authorized' do
      let(:user) { create(:user, :active) }
      before { get :index }

      it 'redirects to dashboard root' do
        expect(response).to redirect_to dashboard_root_path
      end
    end

File: spec/controllers/web/dashboard/users_controller_spec.rb (hunk 2 above)
Comment (reabiliti) [reply spec/controllers/web/dashboard/users_controller_spec.rb]: fix

File: spec/controllers/web/dashboard/users_controller_spec.rb (hunk 3 above)
Comment (reabiliti) [reply spec/controllers/web/dashboard/users_controller_spec.rb]: fix

File: spec/controllers/web/dashboard/users_controller_spec.rb (hunk 4 above)
Comment (reabiliti) [reply spec/controllers/web/dashboard/users_controller_spec.rb]: fix

File: spec/controllers/web/dashboard/users_controller_spec.rb (hunk 5 above)
Comment (reabiliti) [reply spec/controllers/web/dashboard/users_controller_spec.rb]: fix

File: spec/controllers/web/dashboard/users_controller_spec.rb (hunk 6 above)
Comment (reabiliti) [reply spec/controllers/web/dashboard/users_controller_spec.rb]: fix

File: app/operations/ops/developer/add_role.rb (hunk 7 above)
Comment (reabiliti) [reply app/operations/ops/developer/add_role.rb]: fix

File: spec/factories/users.rb (hunk 8 above)
Comment (reabiliti) [reply spec/factories/users.rb]: fix

File: spec/factories/users.rb (hunk 9 above)
Comment (reabiliti) [reply spec/factories/users.rb]: deleted

File: app/operations/ops/developer/disable.rb [hunk 18]
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+module Ops
+  module Developer
+    # When the candidate is approved we run a bunch of tasks.
```
[the remark this answers is not in the document]
Comment (reabiliti) [reply app/operations/ops/developer/disable.rb]: Ok, then I'll leave a comment about this in the file

File: app/operations/ops/developer/remove_role.rb [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+module Ops
+  module Developer
+    # This remove role to user
+    class RemoveRole < BaseOperation
+      step :remove_role!
+
+      private
+
+      def remove_role!(_ctx, user:, role:, size:, **)
+        user.remove_role role if size > 1
```
[the remark this answers is not in the document]
Comment (reabiliti) [reply app/operations/ops/developer/remove_role.rb]: fix

File: app/views/layouts/dashboard.html.haml [hunk 20]
```diff
@@ -1,78 +1,84 @@
...
             %ul.nav.flex-column
               %li.nav-item
-                %a.nav-link.active{:href => "#"}
-                  %span{"data-feather" => "home"}
-                  Dashboard
-                  %span.sr-only (current)
+                %a.nav-link.active{ href: "#" }
+                  %span{ "data-feather": "home" }
+                    Dashboard
+                  %span{ class: 'sr-only' }
```
[the remark this answers is not in the document]
Comment (reabiliti) [reply app/views/layouts/dashboard.html.haml]: fix

File: app/views/layouts/dashboard.html.haml [hunk 21]
```diff
@@ -81,16 +87,17 @@
     Bootstrap core JavaScript
     \==================================================
   / Placed at the end of the document so the pages load faster
-  %script{:crossorigin => "anonymous", :integrity => "sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN", :src => "https://code.jquery.com/jquery-3.2.1.slim.min.js"}
+  %script{ crossorigin: "anonymous", integrity: "sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN",
```
[the remark this answers is not in the document]
Comment (reabiliti) [reply app/views/layouts/dashboard.html.haml]: fix

File: app/controllers/web/dashboard/base_controller.rb [hunk 22]
```diff
@@ -23,6 +23,16 @@ def check_authorization
       def redirect_to_dashboard_root
         redirect_to dashboard_root_url
       end
+
+      def redirect_to_dashboard_users
+        flash[:danger] = t('dashboard.users.access.deny')
+        redirect_to dashboard_users_path
+      end
+
+      def redirect_to_dashboard_user
```
Comment (Mehonoshin) [inline app/controllers/web/dashboard/base_controller.rb]: Этот метод по идее будет использоваться только в UserManagement, поэтому давай его туда и передвинем

File: app/controllers/web/dashboard/users_controller.rb [hunk 23]
```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+module Web
+  module Dashboard
+    # Assigns the role to the user
+    class UsersController < BaseController
+      before_action :user_find, except: :index
+      before_action :authorize_role
+      rescue_from Pundit::NotAuthorizedError, with: :redirect_to_dashboard_users
+      rescue_from CustomErrors::LastRoleError, with: :redirect_to_dashboard_user
```
Comment (Mehonoshin) [inline app/controllers/web/dashboard/users_controller.rb]: `CustomErrors` - quite unclear name
Let's namespace this exception same as the corresponding operation `Ops::Developer::UnassignRole::LastRoleError`

File: app/controllers/web/dashboard/users_controller.rb [hunk 24]
```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+module Web
+  module Dashboard
+    # Assigns the role to the user
+    class UsersController < BaseController
+      before_action :user_find, except: :index
+      before_action :authorize_role
+      rescue_from Pundit::NotAuthorizedError, with: :redirect_to_dashboard_users
```
Comment (Mehonoshin) [inline app/controllers/web/dashboard/users_controller.rb]: For simplicity let's move that to `Dashboard::BaseController` and always redirect to `dashboard_root_url`

File: app/errors/custom_errors.rb [hunk 25]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+module CustomErrors
```
Comment (Mehonoshin) [inline app/errors/custom_errors.rb]: This exception let semantically move to `app/operations/ops/exceptions` for example.

File: app/models/user.rb [hunk 26]
```diff
@@ -25,6 +25,8 @@ class User < ApplicationRecord
 
   accepts_nested_attributes_for :authentications
 
+  scope :active_or_disabled, -> { where(state: %w[active disabled]).order(email: :asc) }
```
Comment (Mehonoshin) [inline app/models/user.rb]: I'd prefer reverse order, to display most recent users first.

File: app/operations/ops/developer/disable.rb [hunk 27]
```diff
@@ -0,0 +1,30 @@
...
+      def change_state!(_ctx, user:, **)
+        user.disable!
+      end
+
+      # def send_notifications!(_ctx, user:, **)
+      # TODO: notify user about starting onboarding
+      # by email
+      # end
+
+      # def start_offboarding!(_ctx, user:, **)
```
Comment (Mehonoshin) [inline app/operations/ops/developer/disable.rb]: We can remove this and create a ticket for that at ZenHub.
It'll make the codebase more clear

File: app/operations/ops/developer/remove_role.rb [hunk 28]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+module Ops
+  module Developer
+    # Removes the role from user
+    class RemoveRole < BaseOperation
+      step :remove_role!
```
Comment (Mehonoshin) [inline app/operations/ops/developer/remove_role.rb]: You can even embed the exception class inside this class.
It'll be the most simple way.

File: app/views/web/dashboard/users/index.html.haml [hunk 29]
```diff
@@ -0,0 +1,30 @@
...
+              - user.roles.each do |role|
+                %li
+                  = role.name
+          %td
+            - if current_user.has_role?(:stuff)
+              - if user.active?
+                = link_to user.state, delete_dashboard_user_path(user), method: :put,
+                                                                        class: "btn btn-success",
+                                                                        data: { confirm: 'Are you sure to disable?' }
+              - else
```
Comment (Mehonoshin) [inline app/views/web/dashboard/users/index.html.haml]: we should also consider the `screening_completed` state.
At this state, we should display button that leads to Applicants page `#show` method, where we can review applicant's quiz results.

File: app/views/web/dashboard/users/show.html.haml [hunk 30]
```diff
@@ -0,0 +1,43 @@
+%h2
+  User id
+  = @user.id
+.table-responsive
+  %table.table.table-striped.table-sm
+    %tr
+      %td
+        %b
+          email
```
Comment (Mehonoshin) [inline app/views/web/dashboard/users/show.html.haml]: Capitalization

File: config/locales/dashboard/en.yml (hunk 17 above)
[the remark this answers is not in the document]
Comment (Mehonoshin) [reply config/locales/dashboard/en.yml]: 👍

File: db/seeds.rb [hunk 31]
```diff
@@ -6,3 +6,7 @@
 tasks.each do |task_attributes|
   Developer::TestTask.find_or_create_by!(task_attributes)
 end
+
+9.times do |i|
+  User.create!(email: "test0#{i}@gmail.com", name: "test0#{i}", state: 'disabled')
```
Comment (Mehonoshin) [inline db/seeds.rb]: Lets `User.create!(email: "test0#{i}@gmail.com", name: "test0#{i}", state: ['disabled', 'activated'].sample)`
to get randomized states

File: spec/controllers/web/dashboard/users_controller_spec.rb [hunk 32]
```diff
@@ -0,0 +1,182 @@
...
+        expect(response).to redirect_to '/'
+      end
+    end
+
+    context 'not authorized' do
+      let(:user) { create(:user, :active) }
+      before { get :index }
+
+      it 'redirect to dashboard root' do
+        expect(response).to redirect_to dashboard_root_path
```
Comment (Mehonoshin) [inline spec/controllers/web/dashboard/users_controller_spec.rb]: An unauthorized user should be redirected to `root_landing_page`.
He is not able to access dashboard.
Please check this behaviour at local environment.

File: app/controllers/web/dashboard/users_controller.rb (hunk 23 above)
Comment (Mehonoshin) [reply app/controllers/web/dashboard/users_controller.rb]: You can even handle this exception at `remove_role` method, without global filtering.

File: config/routes.rb [hunk 33]
```diff
@@ -30,6 +30,14 @@
           put :reject
         end
       end
+      resources :users, only: %i[index show] do
+        member do
+          put :activate
+          put :delete
```
Comment (Mehonoshin) [inline config/routes.rb]: it should be `delete :deactivate`, you got me wrong :)

File: config/routes.rb [hunk 34]
```diff
@@ -30,6 +30,14 @@
...
+          put :delete_role
```
Comment (Mehonoshin) [inline config/routes.rb]: `delete :remove_role`

File: app/controllers/web/dashboard/base_controller.rb (hunk 22 above)
Comment (reabiliti) [reply app/controllers/web/dashboard/base_controller.rb]: fix

File: app/controllers/web/dashboard/users_controller.rb (hunk 23 above)
Comment (reabiliti) [reply app/controllers/web/dashboard/users_controller.rb]: fix

File: app/controllers/web/dashboard/users_controller.rb (hunk 24 above)
Comment (reabiliti) [reply app/controllers/web/dashboard/users_controller.rb]: fix

File: app/models/user.rb (hunk 26 above)
Comment (reabiliti) [reply app/models/user.rb]: fix

File: app/operations/ops/developer/disable.rb (hunk 27 above)
Comment (reabiliti) [reply app/operations/ops/developer/disable.rb]: create issue

**Pull request merged**